libraries/chain/fork_database.cpp

```diff
@@ -712,7 +670,33 @@ namespace eosio::chain {
    }
 
    void fork_database::close() {
-      apply<void>([&](auto& forkdb) { forkdb.close(data_dir / config::forkdb_filename); });
+      auto fork_db_file {data_dir / config::forkdb_filename};
+      bool legacy_valid  = fork_db_l && fork_db_l->in_valid_state();
+      bool savanna_valid = fork_db_s && fork_db_s->in_valid_state();
+
+      // check that fork_dbs are in a consistent state
+      if (!legacy_valid && !savanna_valid) {
+         elog( "fork_database is in a bad state when closing; not writing out '${filename}'", ("filename", fork_db_file) );
```

Comment:
You can log values of `legacy_valid` and `savanna_valid` so it is easier to tell which one causing the problem when trouble shooting.

Reply:
done